File: Notes/Computer Science/Notebook/python/sieve.py
isprime=[1]*100000001
primes=[]
spf=[0]*100000001
order=[0]*100000001
def sieve(a):
	isprime[0]=0
	isprime[1]=0
	for b in range(2,a+1):
		if isprime[b]:
			primes.append(b)
			spf[b]=b
		c=len(primes)
		d=0
		while d<c and b*primes[d]<=a and primes[d]<=spf[b]:
			isprime[b*primes[d]]=0
			spf[b*primes[d]]=primes[d]
			d+=1
	for b in range(2,a+1):
		if isprime[b]:
			order[b]=1
		elif (int(b/spf[b]))%spf[b]!=0:
			order[b]=1
		else:
			order[b]=1+order[int(b/spf[b])]
	return 0
def primefactorsof(a):
	a=int(a)
	av=[]
	while a!=1:
		if len(av)==0 or spf[int(a)]!=av[int(len(av)-1)]:
			av.append(spf[int(a)])
		a/=spf[int(a)]
	return av
def primefactorisationof(a):
	a=int(a)
	avp=[]
	while a!=1:
		if len(avp)==0 or spf[int(a)]!=avp[int(len(avp)-1)][0]:
			avp.append([spf[int(a)],1])
		elif spf[int(a)]==avp[int(len(avp)-1)][0]:
			avp[int(len(avp)-1)][1]+=1
		a/=spf[int(a)]
	return avp
def factorsof(a):
	av=[]
	avp=primefactorisationof(a)
	while len(avp)<15:
		avp.append([1,0])
	avv=[]
	for a1 in range(15):
		bv=[]
		ccc=1
		for a2 in range(1+avp[a1][1]):
			bv.append(ccc)
			ccc*=avp[a1][0]
		avv.append(bv)
	for a1 in range(1+avp[0][1]):
		for a2 in range(1+avp[1][1]):
			for a3 in range(1+avp[2][1]):
				for a4 in range(1+avp[3][1]):
					for a5 in range(1+avp[4][1]):
						for a6 in range(1+avp[5][1]):
							for a7 in range(1+avp[6][1]):
								for a8 in range(1+avp[7][1]):
									for a9 in range(1+avp[8][1]):
										for a10 in range(1+avp[9][1]):
											for a11 in range(1+avp[10][1]):
												for a12 in range(1+avp[11][1]):
													for a13 in range(1+avp[12][1]):
														for a14 in range(1+avp[13][1]):
															for a15 in range(1+avp[14][1]):
																av.append(avv[0][a1]*avv[1][a2]*avv[2][a3]*avv[3][a4]*avv[4][a5]*avv[5][a6]*avv[6][a7]*avv[7][a8]*avv[8][a9]*avv[9][a10]*avv[10][a11]*avv[11][a12]*avv[12][a13]*avv[13][a14]*avv[14][a15])
	av.sort()
	return av

#sieve(100000000)
#print(factorsof(1000))
#Runtime on Colab is ~ 150s
#Can store locally to avoid.

File: Notes/Computer Science/Notebook/math/sieve.cpp
// Better In Loop
Z az(){
	I maximum=10000000; // SSRS_
	V spf(maximum+1,-1);
	V primes;
	for(a=2;a<=maximum;a++){
		if(spf[a]==-1){
			primes.add(a);
			for(b=a;b<=maximum;b+=a){
				if(spf[b]==-1)spf[b]=a;
			}
		}
	}

	in>>a;
	for(b=0;b<a;b++){
		in>>c>>d;
		if(gcd(c,d)>1){
			out<<0<<nl;
		}
		else if(c==d-1){
			out<<-1<<nl;
		}
		else{
			e=d-c;
			m=INT_MAX;
			while(e>1){
				f=spf[e];
				m=min(m,(f-(c%f)));
				e/=f;
			}
			out<<m<<nl;
		}
	}
}

// ----------

VB isprime(1000001,true);
VL primes;
VL spf(1000001);
VL order(1000001);
Z sieve(L al){
    isprime[0]=isprime[1]=false;
    I c;
    for(L b=2;b<=al;b++){
        if(isprime[b]){
            primes.add(b);
            spf[b]=b;
        }
        c=sz(primes);
        for(L d=0;d<c && b*primes[d]<=al && primes[d]<=spf[b];d++){
            isprime[b*primes[d]]=false;
            spf[b*primes[d]]=primes[d];
        }
    }
	for(L b=2;b<=al;b++){
		if(isprime[b]){
			order[b]=1;
		}
		else if((b/spf[b])%spf[b]!=0){
			order[b]=1;
		}
		else{
			order[b]=1+order[b/spf[b]];
		}
	}
}
 
V primefactorsof(I a){
	V av;
	while(a!=1){
		if(sz(av)==0 or spf[a]!=av[sz(av)-1])av.add(spf[a]);
		a/=spf[a];
	}
	return av;
}
 
VP primefactorisationof(I a){
	VP avp;
	while(a!=1){
		if(sz(avp)==0 or spf[a]!=avp[sz(avp)-1].first)avp.add({spf[a],1});
		else if(spf[a]==avp[sz(avp)-1].first)avp[sz(avp)-1].second++;
		a/=spf[a];
	}
	return avp;
}

V factorsof(I n){
	VP avp=primefactorisationof(n);
	V av,bv;
	VV avv;
	I a0,a1,a2,a3,a4,a5,a6,a7,a8,a9,a10,a11,a12,a13,a14;
	while(sz(avp)<15)avp.add({1,0});
	for(a0=0;a0<15;a0++){
		bv=V(0);
		I c=1;
		for(a1=0;a1<=avp[a0].second;a1++){
			bv.add(c);
			c*=avp[a0].first;
		}
		avv.add(bv);
	}
	for(a0=0;a0<=avp[0].second;a0++){
	for(a1=0;a1<=avp[1].second;a1++){
	for(a2=0;a2<=avp[2].second;a2++){
	for(a3=0;a3<=avp[3].second;a3++){
	for(a4=0;a4<=avp[4].second;a4++){
	for(a5=0;a5<=avp[5].second;a5++){
	for(a6=0;a6<=avp[6].second;a6++){
	for(a7=0;a7<=avp[7].second;a7++){
	for(a8=0;a8<=avp[8].second;a8++){
	for(a9=0;a9<=avp[9].second;a9++){
	for(a10=0;a10<=avp[10].second;a10++){
	for(a11=0;a11<=avp[11].second;a11++){
	for(a12=0;a12<=avp[12].second;a12++){
	for(a13=0;a13<=avp[13].second;a13++){
	for(a14=0;a14<=avp[14].second;a14++){
		av.add(avv[0][a0]*avv[1][a1]*avv[2][a2]*avv[3][a3]*avv[4][a4]*avv[5][a5]*avv[6][a6]*avv[7][a7]*avv[8][a8]*avv[9][a9]*avv[10][a10]*avv[11][a11]*avv[12][a12]*avv[13][a13]*avv[14][a14]);
	}
	}
	}
	}
	}
	}
	}
	}
	}
	}
	}
	}
	}
	}
	}
	sort(all(av));
	return av;
}